File: client/src/app/service/board-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { CreateBoardDto } from './dto/create-board-dto';
import { ResultDto } from './dto/result-dto';
import { CellIdDto } from './dto/cell-id-dto';
import { BoardIdDto } from './dto/board-id-dto';
import { BoardDataDto } from './dto/board-data-dto';
import { BoardDto } from './dto/board-dto';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class BoardService {

  constructor(private http: HttpClient) { }

  create(dto: CreateBoardDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/create", dto);
  }

  pause(dto: BoardIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/pause", dto);
  }

  resume(dto: BoardIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/resume", dto);
  }

  find(): Observable<ResultDto<BoardDataDto[]>> {
    return this.http.get<ResultDto<BoardDataDto[]>>("/find");
  }

  delete(dto: BoardIdDto): Observable<ResultDto<any>> {
    return this.http.post<ResultDto<any>>("/delete", dto);
  }

  click(dto: CellIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/click", dto);
  }

  redFlag(dto: CellIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/redFlag", dto);
  }

  questionMark(dto: CellIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/questionMark", dto);
  }

  initial(dto: CellIdDto): Observable<ResultDto<BoardDto>> {
    return this.http.post<ResultDto<BoardDto>>("/initial", dto);
  }
}

File: core/src/main/java/com/minesweeper/service/api/BoardService.java
package com.minesweeper.service.api;

import com.minesweeper.common.api.dto.BoardDto;
import com.minesweeper.service.api.dto.*;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * This is not used, right now, but it could be
 * helpful when creating a client (for example, Retrofit).
 * Of course it should be put into a separated project.
 */
public interface BoardService {
    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> create(@Nonnull CreateBoardDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> pause(@Nonnull BoardIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> resume(@Nonnull BoardIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<List<BoardDataDto>>> find();

    @Nonnull
    CompletableFuture<ResultDto<Void>> delete(@Nonnull BoardIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> click(@Nonnull CellIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> redFlag(@Nonnull CellIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> questionMark(@Nonnull CellIdDto dto);

    @Nonnull
    CompletableFuture<ResultDto<BoardDto>> initial(@Nonnull CellIdDto dto);
}
